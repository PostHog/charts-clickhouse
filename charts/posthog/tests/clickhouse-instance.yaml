suite: ClickHouse instance
templates:
  - templates/clickhouse_instance.yaml

tests:
  - it: should be empty if clickhouse.enabled is set to false
    set:
      clickhouse.enabled: false
    asserts:
      - hasDocuments:
          count: 0

  - it: the manifest should match the snapshot when using default values
    set:
      clickhouse.enabled: true
    asserts:
      - hasDocuments:
          count: 1
      - matchSnapshot: {}

  - it: should have the correct apiVersion
    set:
      clickhouse.enabled: true
    asserts:
      - hasDocuments:
          count: 1
      - isAPIVersion:
          of: clickhouse.altinity.com/v1

  - it: should be the correct kind
    set:
      clickhouse.enabled: true
    asserts:
      - hasDocuments:
          count: 1
      - isKind:
          of: ClickHouseInstallation

  - it: spec.configuration.cluster[0].templates volumeClaimTemplate should not exist if persistence is disabled
    set:
      clickhouse.persistence.enabled: false
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.configuration.clusters[0].templates
          value:
            podTemplate: pod-template
            clusterServiceTemplate: service-template

  - it: spec.configuration.cluster[0].templates volumeClaimTemplate should not exist if persistence is enabled and clickhouse.persistence.existingClaim is set
    set:
      clickhouse.persistence.enabled: false
      clickhouse.persistence.existingClaim: pvc-claim
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.configuration.clusters[0].templates
          value:
            podTemplate: pod-template
            clusterServiceTemplate: service-template

  - it: spec.configuration.cluster[0].templates volumeClaimTemplate should exist if persistence is enabled
    set:
      clickhouse.persistence.enabled: true
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.configuration.clusters[0].templates
          value:
            podTemplate: pod-template
            clusterServiceTemplate: service-template
            dataVolumeClaimTemplate: data-volumeclaim-template


  - it: pod-template spec.volumes should not exist if persistence is disabled
    set:
      clickhouse.persistence.enabled: false
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.templates.podTemplates[0].spec.volumes
          value: null

  - it: pod-template spec.volumes should have 'data-volumeclaim-template' if persistence is enabled
    set:
      clickhouse.persistence.enabled: true
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.templates.podTemplates[0].spec.volumes
          value:
            - name: data-volumeclaim-template
              persistentVolumeClaim:
                claimName: data-volumeclaim-template

  - it: pod-template spec.volumes should have 'existing-volumeclaim' if clickhouse.persistence.existingClaim is set
    set:
      clickhouse.persistence.enabled: true
      clickhouse.persistence.existingClaim: pvc-claim
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.templates.podTemplates[0].spec.volumes
          value:
            - name: existing-volumeclaim
              persistentVolumeClaim:
                claimName: pvc-claim

  - it: pod-template spec.containers[0].volumeMounts should not exist if persistence is disabled
    set:
      clickhouse.persistence.enabled: false
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.templates.podTemplates[0].spec.containers[0].volumeMounts
          value: null

  - it: pod-template spec.containers[0].volumeMounts should exist if persistence is enabled
    set:
      clickhouse.persistence.enabled: enabled
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.templates.podTemplates[0].spec.containers[0].volumeMounts
          value:
            - mountPath: /var/lib/clickhouse
              name: data-volumeclaim-template

  - it: pod-template spec.containers[0].volumeMounts should have 'existing-volumeclaim' if clickhouse.persistence.existingClaim is set
    set:
      clickhouse.persistence.enabled: enabled
      clickhouse.persistence.existingClaim: pvc-claim
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.templates.podTemplates[0].spec.containers[0].volumeMounts
          value:
            - mountPath: /var/lib/clickhouse
              name: existing-volumeclaim

  - it: nodeSelector override via 'clickhouse.nodeSelector' works
    set:
      clickhouse.nodeSelector:
        diskType: ssd
        nodeType: fast
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.templates.podTemplates[0].spec.nodeSelector
          value:
            diskType: ssd
            nodeType: fast

  - it: volumeClaimTemplates shouldn't exit if clickhouse.persistence.enabled is false
    set:
      clickhouse.persistence.enabled: false
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.templates.volumeClaimTemplates
          value: null

  - it: volumeClaimTemplates shouldn't exit if clickhouse.persistence.enabled is true and clickhouse.persistence.existingClaim is set
    set:
      clickhouse.persistence.enabled: false
      clickhouse.persistence.existingClaim: pvc-claim
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.templates.volumeClaimTemplates
          value: null

  - it: The storageClassName value should be null using the default values
    set:
      clickhouse.persistence.enabled: true
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.templates.volumeClaimTemplates[0].spec.storageClassName
          value: null

  - it: The storageClassName override should work
    set:
      clickhouse.persistence.enabled: true
      clickhouse.persistence.storageClass: customStorageClass
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.templates.volumeClaimTemplates[0].spec.storageClassName
          value: customStorageClass

  - it: default profiles configuration should be correct
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.configuration.profiles
          value:
            default/allow_experimental_window_functions: "1"

  - it: handles custom profiles configuration
    set:
      clickhouse.profiles.default/max_execution_time: "90"
      clickhouse.profiles.default/max_memory_usage: "20000000000"
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.configuration.profiles
          value:
            default/allow_experimental_window_functions: "1"
            default/max_execution_time: "90"
            default/max_memory_usage: "20000000000"


  - it: handles custom profiles overriding defaults
    set:
      clickhouse.profiles.default/allow_experimental_window_functions: "0"
    asserts:
      - hasDocuments:
          count: 1
      - equal:
          path: spec.configuration.profiles
          value:
            default/allow_experimental_window_functions: "0"

  - it: handles custom users and passwords
    set:
      clickhouse.user: someuser
      clickhouse.password: somepass
    asserts:
      - equal:
          path: spec.configuration.users
          value:
            someuser/password: somepass
            someuser/networks/ip: "0.0.0.0/0"
            someuser/profile: default
            someuser/quota: default

  - it: sets default clickhouse-server image
    asserts:
      - equal:
          path: spec.templates.podTemplates[0].spec.containers[0].image
          value: "yandex/clickhouse-server:21.6.5"

  - it: allows modifying clickhouse-server version
    set:
      clickhouse.image.repository: altinity/clickhouse-server
      clickhouse.image.tag: 22.1
    asserts:
      - equal:
          path: spec.templates.podTemplates[0].spec.containers[0].image
          value: "altinity/clickhouse-server:22.1"

  - it: allows overriding shard counts
    set:
      clickhouse.layout.shardsCount: 3
      clickhouse.layout.replicasCount: 2
    asserts:
      - equal:
          path: spec.configuration.clusters[0].layout
          value:
            shardsCount: 3
            replicasCount: 2

  - it: allows passing custom sharding settings
    set:
      clickhouse.layout:
        shardsCount: null
        replicasCount: null
        # From https://github.com/Altinity/clickhouse-operator/blob/master/docs/custom_resource_explained.md
        shards:
          - name: shard0
            replicasCount: 3
            weight: 1
            internalReplication: Disabled
            templates:
              podTemplate: clickhouse-v18.16.1
              dataVolumeClaimTemplate: default-volume-claim
              logVolumeClaimTemplate: default-volume-claim

          - name: shard1
            templates:
              podTemplate: clickhouse-v18.16.1
              dataVolumeClaimTemplate: default-volume-claim
              logVolumeClaimTemplate: default-volume-claim
            replicas:
              - name: replica0
              - name: replica1
              - name: replica2
    asserts:
      - matchSnapshot:
          path: spec.configuration.clusters[0].layout
