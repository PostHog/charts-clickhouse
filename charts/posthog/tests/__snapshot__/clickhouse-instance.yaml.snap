allows passing custom sharding settings:
  1: |
    shards:
    - internalReplication: Disabled
      name: shard0
      replicasCount: 3
      templates:
        dataVolumeClaimTemplate: default-volume-claim
        logVolumeClaimTemplate: default-volume-claim
        podTemplate: clickhouse-v18.16.1
      weight: 1
    - name: shard1
      replicas:
      - name: replica0
      - name: replica1
      - name: replica2
      templates:
        dataVolumeClaimTemplate: default-volume-claim
        logVolumeClaimTemplate: default-volume-claim
        podTemplate: clickhouse-v18.16.1
the manifest should match the snapshot when using default values:
  1: |
    apiVersion: clickhouse.altinity.com/v1
    kind: ClickHouseInstallation
    metadata:
      name: posthog
    spec:
      configuration:
        clusters:
        - layout:
            replicasCount: 1
            shardsCount: 1
          name: posthog
          templates:
            clusterServiceTemplate: service-template
            dataVolumeClaimTemplate: data-volumeclaim-template
            podTemplate: pod-template
        files:
          events.proto: |
            syntax = "proto3";
            message Event {
              string uuid = 1;
              string event = 2;
              string properties = 3;
              string timestamp = 4;
              uint64 team_id = 5;
              string distinct_id = 6;
              string created_at = 7;
              string elements_chain = 8;
            }
        profiles:
          default/allow_experimental_window_functions: "1"
          default/allow_nondeterministic_mutations: "1"
        settings:
          default_database: posthog
          format_schema_path: /etc/clickhouse-server/config.d/
        users:
          admin/networks/ip:
          - 10.0.0.0/8
          - 172.16.0.0/12
          - 192.168.0.0/16
          admin/password: a1f31e03-c88e-4ca6-a2df-ad49183d15d9
          admin/profile: default
          admin/quota: default
        zookeeper:
          nodes:
          - host: RELEASE-NAME-posthog-zookeeper
            port: 2181
      templates:
        podTemplates:
        - name: pod-template
          spec:
            containers:
            - command:
              - /bin/bash
              - -c
              - /usr/bin/clickhouse-server --config-file=/etc/clickhouse-server/config.xml
              image: clickhouse/clickhouse-server:23.4
              name: clickhouse
              ports:
              - containerPort: 8123
                name: http
              - containerPort: 9000
                name: client
              - containerPort: 9009
                name: interserver
              volumeMounts:
              - mountPath: /var/lib/clickhouse
                name: data-volumeclaim-template
            securityContext:
              fsGroup: 101
              runAsGroup: 101
              runAsUser: 101
            volumes:
            - name: data-volumeclaim-template
              persistentVolumeClaim:
                claimName: data-volumeclaim-template
        serviceTemplates:
        - generateName: clickhouse-RELEASE-NAME
          name: service-template
          spec:
            ports:
            - name: http
              port: 8123
            - name: tcp
              port: 9000
            type: ClusterIP
        volumeClaimTemplates:
        - name: data-volumeclaim-template
          spec:
            accessModes:
            - ReadWriteOnce
            resources:
              requests:
                storage: 20Gi
